subsys/fs/nvs/nvs.c

```diff
@@ -6,104 +6,492 @@
+ * fs->write_block_size. _nvs_flash_move does a verification of the copy.
+ */
+int _nvs_flash_block_move(struct nvs_fs *fs, off_t offset, size_t len)
+{
+	int rc, byte_to_copy, block_size;
+	u8_t buf[max(NVS_BLOCK_SIZE, fs->write_block_size)];
+	off_t write_location;
+
+	write_location = fs->write_location;
+
+	block_size = sizeof(buf);
+
+	while (len > 0) {
+		byte_to_copy =	min(block_size, len);
+		rc = _nvs_flash_read(fs, offset, buf, byte_to_copy);
+
+		if (rc) {
+			return rc;
+		}
+		rc = _nvs_flash_write(fs, write_location, buf,
+				      byte_to_copy);
+		if (rc) {
+			return rc;
+		}
+		rc = _nvs_flash_block_cmp(fs, write_location, buf,
+					  byte_to_copy);
+		if (rc) {
+			return rc;
+		}
+		len -= byte_to_copy;
+		offset += byte_to_copy;
+		write_location += byte_to_copy;
+	}
+	/* update fs->write_location when all writes are successful */
+
+	fs->write_location = write_location;
+
+	return 0;
+}
+
+/* flash write and verify */
+int _nvs_flash_write_ver(struct nvs_fs *fs, off_t offset, const void *data,
+			 size_t len)
 {
-	struct _nvs_sector_hdr sector_hdr;
-	u16_t sector_hdr_len;
 	int rc;
 
-	/* just to be sure, align offset to sector_size */
-	offset &= ~(fs->sector_size - 1);
-	fs->sector_id += 1;
-	rc = _nvs_sector_hdr_get(fs, offset, &sector_hdr);
+	rc = _nvs_flash_write(fs, offset, data, len);
 	if (rc) {
-		SYS_LOG_DBG("Failed reading sector hdr");
 		return rc;
 	}
-	if (sector_hdr.fd_magic != 0xFFFFFFFF) {
-		SYS_LOG_DBG("Initializing non empty sector");
-		return -EIO;
+	rc = _nvs_flash_block_cmp(fs, offset, data, len);
+	if (rc) {
+		return rc;
 	}
-	sector_hdr.fd_magic = fs->magic;
-	sector_hdr.fd_id = fs->sector_id;
-	sector_hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr));
-	rc = nvs_flash_write(fs, offset, &sector_hdr, sizeof(sector_hdr));
+	return 0;
+}
+
+/* store an entry in flash at location offset */
+int _nvs_flash_wrt_entry(struct nvs_fs *fs, u16_t id, const void *data,
+			 size_t len)
+{
+	int rc;
+	off_t offset;
+	off_t fs_size = fs->sector_count * fs->sector_size;
+	size_t padding, flash_len, length_cnt, byte_cnt;
+	u8_t lenstr[2], ilenstr[2], buf[max(fs->write_block_size, 4)];
+
+	offset = fs->write_location;
+
+	/* create the length string to be written at start and end */
+	length_cnt = _nvs_put_length(lenstr, len);
+	if (length_cnt == 1) {
+		ilenstr[0] = lenstr[0];
+	} else {
+		ilenstr[0] = lenstr[1];
+		ilenstr[1] = lenstr[0];
+	}
+
+	/* calculate the entry length in flash */
+	flash_len = _nvs_flash_size(fs, len);
+
+	/* calculate the padding length */
+	padding = flash_len - len - sizeof(u16_t) - 2*length_cnt;
+
+	memcpy(buf, lenstr, (size_t) length_cnt);
+	memcpy(buf + length_cnt, (u8_t *)&id, 2);
+	byte_cnt = length_cnt + 2;
+
+	if (byte_cnt >= fs->write_block_size) {
+		rc = _nvs_flash_write_ver(fs, offset, buf,
+					 byte_cnt);
+		if (rc) {
+			/* flash write or verify error */
+			goto err;
+		}
+		offset += byte_cnt;
+		byte_cnt = 0;
+	}
```

Comment:
All the cases in this function looks like error prone. For example if `write_block_size` is 2 bytes and the data length to be written below 128 bytes, `byte_cnt` will be 3. This mean the first call to `_nvs_flash_write_ver` will try to write 3 bytes, which is the wrong number for a 2-byte write block size.

Reply:
This is an error, I will review this function once again to make it less error prone.